src/main/scala/io/iohk/ethereum/crypto/ECDSASignature.scala

```diff
@@ -95,22 +39,72 @@ object ECDSASignature {
+    * v = CHAIN_ID * 2 + 35 or v = CHAIN_ID * 2 + 36
+    */
+  private def getRecoveredPointSign(pointSign: Byte, chainId: Option[Byte]): Option[Int] = {
+    chainId match {
+      case Some(id) =>
+        if (pointSign == negativePointSign || pointSign == (id * 2 + newNegativePointSign).toByte) {
+          Some(negativePointSign)
+        } else if (pointSign == positivePointSign || pointSign == (id * 2 + newPositivePointSign).toByte) {
+          Some(positivePointSign)
+        } else {
+          None
+        }
+      case None => Some(pointSign)
+    }
+  }
 
-case class ECDSASignature(r: BigInteger, s: BigInteger, v: Byte) {
+  private def canonicalise(s: BigInteger): BigInteger = {
+    val halfCurveOrder = curveParams.getN.shiftRight(1)
+    if (s.compareTo(halfCurveOrder) > 0) curve.getN.subtract(s)
+    else s
+  }
+
+  private def calculateV(r: BigInteger, s: BigInteger, key: AsymmetricCipherKeyPair, message: Array[Byte]): Option[Byte] = {
+    val pubKey = key.getPublic.asInstanceOf[ECPublicKeyParameters].getQ.getEncoded(false)
+    val recIdOpt = Seq(positivePointSign, negativePointSign).find { i =>
+      recoverPubBytes(r, s, i, None, message).exists(java.util.Arrays.equals(_, pubKey))
+    }
+    recIdOpt
+  }
 
-  import ECDSASignature._
+  private def recoverPubBytes(r: BigInteger, s: BigInteger, recId: Byte, chainId: Option[Byte], message: Array[Byte]): Option[Array[Byte]] = {
+    val order = curve.getCurve.getOrder
+    val xCoordinate = r //ignore case when x = r + order because it is negligibly improbable
+    val curveFp = curve.getCurve.asInstanceOf[ECCurve.Fp]
+    val prime = curveFp.getQ
+
+    getRecoveredPointSign(recId, chainId).flatMap { recovery =>
+      if (xCoordinate.compareTo(prime) < 0) {
+        val R = constructPoint(xCoordinate, recovery)
+        if (R.multiply(order).isInfinity) {
+          val e = new BigInteger(1, message)
+          val rInv = r.modInverse(order)
+          //Q = r^(-1)(sR - eG)
+          val q = R.multiply(s).subtract(curve.getG.multiply(e)).multiply(rInv)
+          Some(q.getEncoded(false))
+        } else None
+      } else None
+    }
+  }
 
-  lazy val encoded: ByteString = {
-    ByteString(
-      asUnsignedByteArray(r).reverse.padTo(RLength, 0.toByte).reverse ++
-        asUnsignedByteArray(s).reverse.padTo(SLength, 0.toByte).reverse ++
-        Array(recIdFromSignatureV(v)))
+  private def constructPoint(xCoordinate: BigInteger, recId: Int): ECPoint = {
+    val x9 = new X9IntegerConverter
+    val compEnc = x9.integerToBytes(xCoordinate, 1 + x9.getByteLength(curve.getCurve))
+    compEnc(0) = if (recId == ECDSASignature.positivePointSign) 3.toByte else 2.toByte
+    curve.getCurve.decodePoint(compEnc)
   }
 
-  def recoverPubBytes(message: Array[Byte]): Option[Array[Byte]] =
-    ECDSASignature.recoverPubBytes(r, s, recIdFromSignatureV(v), message)
+}
+
+case class ECDSASignature(r: BigInteger, s: BigInteger, v: Byte) {
```

Comment:
I know r,s, and v are conventions but could we have some link or hint to the casual reader that these are not just lazyily named?